### PR TITLE
feat(VDataTable): add row index to item cell slot

### LINKED ISSUE #10646: [Feature Request] VDataTable: add missing index variable in "item.<name>" scoped slot
### Problem to solve
It would be very useful for adding a column with row index, and using index for custom stuff in a cell.

### Proposed solution
Inside the scoped slot "item.\<name>", the props should also contain the "index" property, like in the "item" scoped slot (cf #8199 and #8283).

### PR DESCRIPTION
**Pull request by alicjamusial**
Add row index to item cell slot to be able to introduce rows numeration while using custom slots

## Description

Add a possiblility to use row index in custom items props cell.

## Motivation and Context

Feature was added to allow users to introduce rows numeration easily, even if user is using custom item column props.

fixes #10646

## How Has This Been Tested?

Visually (playground).

## Markup:

<details>

```vue
<template>
  <v-container>
    <v-data-table :headers="tableHeaders" :items="tableData">
      <template v-slot:item.index="{ index }">
        {{ index }}
      </template>
      <template v-slot:item.name="{ item }">
        {{ item.name }}
      </template>
    </v-data-table>
  </v-container>
</template>

<script>
export default {
  data() {
    return {
      tableHeaders: [
          {
            text: 'No',
            sortable: false,
            value: 'index'
          },
          {
            text: 'Name',
            sortable: true,
            value: 'name'
          }
        ],
      tableData: [
          {
            name: 'Data 1'
          },
          {
            name: 'Data 2'
          },
          {
            name: 'Data 3'
          },
        ]
    }
  }
}
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VDataTable/VDataTable.ts [hunk 1]
```diff
@@ -422,7 +422,7 @@ export default VDataIterator.extend({
         this.$createElement('template', { slot: 'row.content' }, [expandedRow]),
       ])
     },
-    genDefaultSimpleRow (item: any, classes: Record<string, boolean> = {}): VNode {
+    genDefaultSimpleRow (item: any, classes: Record<string, boolean> = {}, index: Number = 0): VNode {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VDataTable/VDataTable.ts]:
```suggestion
    genDefaultSimpleRow (item: any, classes: Record<string, boolean> = {}, index = 0): VNode {
```

**Comment by alicjamusial** [comment]
All requested changes done.

**Comment by asterd** [comment]
I think this can be a very good addition in many cases? any chance to integrate into next minor release?

**Comment by jacekkarczmarczyk** [comment]
When items are paginated row numerations starts from 0 on next pages, I don't think that's correct

image

<details>

```vue
<template>
  <v-data-table
    :headers="tableHeaders"
    :items="tableData"
  >
    <template v-slot:item.index="{ index }">{{ index }}</template>
    <template v-slot:item.name="{ item }">{{ item.name }}</template>
  </v-data-table>
</template>

<script>
  export default {
    data () {
      return {
        tableHeaders: [
          {
            text: 'No',
            sortable: false,
            value: 'index',
          },
          {
            text: 'Name',
            sortable: true,
            value: 'name',
          }
        ],
        tableData: [
          { name: 'Data 1' },
          { name: 'Data 2' },
          { name: 'Data 3' },
          { name: 'Data 4' },
          { name: 'Data 5' },
          { name: 'Data 6' },
          { name: 'Data 7' },
        ]
      }
    }
  }
</script>
```

</details>

File: packages/api-generator/src/maps/v-data-table.js [hunk 2]
```diff
@@ -83,6 +83,7 @@ const DataTableItemColumnScopedProps = {
   item: 'any',
   header: 'TableHeader',
   value: 'any',
+  index: 'number',
```
Comment (jacekkarczmarczyk) [inline packages/api-generator/src/maps/v-data-table.js:86]: it should be added also to `DataTableItemScopedProps` and `DataTableExpandedItemScopedProps` (any others?)

**Comment by johnleider** [comment]
Unfortunately this did not make it in for v2.3.

Thank you for your contribution and interest in improving Vuetify. Make sure to join us in the [Discord community](https://community.vuetifyjs.com).

**Pull request closed without merging**